packages/x-tree-view/src/internals/plugins/useTreeViewSelection/useTreeViewSelection.itemPlugin.ts

```diff
@@ -63,8 +63,10 @@ function selectorItemCheckboxStatus(
   return {
     disabled: !isSelectionEnabledForItem,
     visible: isCheckboxSelectionEnabled,
-    indeterminate: (hasSelectedDescendant && hasUnSelectedDescendant) || !hasUnSelectedDescendant,
-    checked: selectorSelectionPropagationRules(state).parents ? hasSelectedDescendant : false,
+    indeterminate: hasSelectedDescendant && hasUnSelectedDescendant,
+    checked: selectorSelectionPropagationRules(state).parents
+      ? hasSelectedDescendant && !hasUnSelectedDescendant
```

Comment:
This is the actual change, adding `!hasUnSelectedDescendant` makes sure that when your checkbox has unselected children, clicking on it will select them.